### PR TITLE
backend-agnostic quantum info module

### PR DESCRIPTION
**Pull request by renatomello**
Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Comment by renatomello** [comment]
This goes together with https://github.com/qiboteam/qibojit/pull/115 and closes #843

**Comment by renatomello** [comment]
This is ready @vodovozovaliza @stavros11

File: src/qibo/gates/channels.py
Comment (stavros11) [inline src/qibo/gates/channels.py:104]: Maybe these comments (and same in the following lines) are not needed.

File: src/qibo/quantum_info/basis.py
Comment (stavros11) [inline src/qibo/quantum_info/basis.py:193]: If you do this `cast` for all backends, so that you don't need to check explicitly if it is tensorflow, does it break something?

File: tests/test_cirq.py
Comment (stavros11) [inline tests/test_cirq.py:161]: What is the problem with the `CupyBackend` here? I believe these tests used to pass with it.

File: src/qibo/quantum_info/metrics.py
Comment (stavros11) [inline src/qibo/quantum_info/metrics.py:36]: I see that this is done in many functions, why is it needed?

**Review by stavros11 — COMMENTED** [review]
Thanks for implementing this. I left some comments below which are mostly questions. My main concern is why you need to skip the Cirq tests for cupy (and some for tensorflow) as I believe they were working before and I think what they are testing is not very related to this PR. If they are failing now, it may be worth checking why.

Other than that it looks good to me. I was thinking whether we could simplify this mechanism of
```py
if backend is None: 
    backend = GlobalBackend()
```
which is extensively used here but also in other places in the code but I am not sure how. Some of the ideas in #761 are worth checking and could help in this direction, but this may take more time so we should not block this PR for it.

File: src/qibo/quantum_info/basis.py
Comment (renatomello) [reply src/qibo/quantum_info/basis.py:193]: It doesn't break but it's redundant. I opted to check if because for bigger systems it might be a slow process to recast to the same backend even though the array is already at the right backend.

File: tests/test_cirq.py
Comment (renatomello) [reply tests/test_cirq.py:161]: There are some `Cirq` fucntions that don't work with `cupy.ndarray` inputs due to the way they're implemented. Same thing for `EagerTensor`s. I even proposed a change to `Cirq` [here](https://github.com/quantumlib/Cirq/pull/6076) to fix some of these issues.

File: src/qibo/quantum_info/metrics.py
Comment (renatomello) [reply src/qibo/quantum_info/metrics.py:36]: Otherwise it returns a 0-dim array with a float inside.

File: src/qibo/gates/channels.py
Comment (renatomello) [reply src/qibo/gates/channels.py:104]: Thanks

**Comment by renatomello** [comment]
> Thanks for implementing this. I left some comments below which are mostly questions. My main concern is why you need to skip the Cirq tests for cupy (and some for tensorflow) as I believe they were working before and I think what they are testing is not very related to this PR. If they are failing now, it may be worth checking why.
> 
> Other than that it looks good to me. I was thinking whether we could simplify this mechanism of
> 
> ```python
> if backend is None: 
>     backend = GlobalBackend()
> ```
> 
> which is extensively used here but also in other places in the code but I am not sure how. Some of the ideas in #761 are worth checking and could help in this direction, but this may take more time so we should not block this PR for it.

I also prefer to leave this to another PR.

File: src/qibo/quantum_info/metrics.py
Comment (renatomello) [inline src/qibo/quantum_info/metrics.py:82]: Here, its non-Hermitian `state - target`.

File: src/qibo/quantum_info/basis.py
Comment (renatomello) [reply src/qibo/quantum_info/basis.py:193]: Changed where arrays are cast to the general backend, and now this is working

**Review by stavros11 — APPROVED** [review]
Thanks for the updates and explanations.

Regarding `backend.cast`, if you see the implementation in the `NumpyBackend`
https://github.com/qiboteam/qibo/blob/2753c4fe4a59cccc8f42f4c23f2ca6cb0ef71bde/src/qibo/backends/numpy.py#L56
if the input `x` is already a numpy array, it will return `x.astype(dtype, copy=False)`, which according to the [numpy documentation](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html#numpy.ndarray.astype) will do nothing if `x` already has type `dtype`. So there should be no performance deterioration even if you use `cast` when is not really needed. Note that `NumbaBackend` inherits `cast` from `NumpyBackend`. `if backend.__class__.__name__` checks is not a blocking issue for me, it just looks a bit annoying because it breaks the symmetry between different backends. One thing to point out though is that if you call `backend.cast` without specifying the `dtype`, it will use the default type (which is `complex128`) and this may create a copy of the array while casting to a different type, or even cast a real array to complex which you may not want.

**Comment by renatomello** [comment]
> Thanks for the updates and explanations.
> 
> Regarding `backend.cast`, if you see the implementation in the `NumpyBackend`
> 
> https://github.com/qiboteam/qibo/blob/2753c4fe4a59cccc8f42f4c23f2ca6cb0ef71bde/src/qibo/backends/numpy.py#L56
> 
> if the input `x` is already a numpy array, it will return `x.astype(dtype, copy=False)`, which according to the [numpy documentation](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.astype.html#numpy.ndarray.astype) will do nothing if `x` already has type `dtype`. So there should be no performance deterioration even if you use `cast` when is not really needed. Note that `NumbaBackend` inherits `cast` from `NumpyBackend`. `if backend.__class__.__name__` checks is not a blocking issue for me, it just looks a bit annoying because it breaks the symmetry between different backends. One thing to point out though is that if you call `backend.cast` without specifying the `dtype`, it will use the default type (which is `complex128`) and this may create a copy of the array while casting to a different type, or even cast a real array to complex which you may not want.

Thanks for that explanation. Regarding `dtype`, from my experience, it is needed because `Tensorflow` does not accept `complex128`.

File: tests/test_gates_channels.py
Comment (renatomello) [reply tests/test_gates_channels.py]: Done

**Comment by scarrazza** [comment]
@renatomello are you sure this will work without pinning the latest qibojit version in the toml?

**Comment by renatomello** [comment]
> @renatomello are you sure this will work without pinning the latest qibojit version in the toml?

`pyproject.toml` is installing `qibojit` from `master`. Since the changes are already merged there, it is supposed to work (e.g. tests are passing here). 

As I understand, the tests failed on main after merge in an issue that seemed unrelated to this PR.

**Pull request merged**
